### PR TITLE
[tune] Remove TrialCheckpoint class

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

The old user-facing TrialCheckpoint class has been deprecated in favor of `ray.ml.Checkpoint` and will be removed with this PR.

The main change in this PR is to delete the old `TrialCheckpoint` class and replace remaining API calls (e.g. `checkpoint.local_path`) with the correct AIR equivalents.

One issue that comes up is that with Ray client usage, checkpoint directories are not available on the local node (the client). Thus, we can't construct `Checkpoint` objects easily. (Previously, the TrialCheckpoint object held a reference to the location, even if it is not locally available). There are ongoing discussions on how to resolve this in the future. For now, we print an error when such a checkpoint is requested.

Depends on #25805

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ArturNiederfahrenhorst — APPROVED** [review]

**Review by Yard1 — APPROVED** [review]

**Comment by krfricke** [comment]
Most release tests are passing: https://buildkite.com/ray-project/release-tests-pr/builds/8843#_

For the failing test (which uses client + checkpoints): https://buildkite.com/ray-project/release-tests-pr/builds/8847

File: python/ray/tune/analysis/experiment_analysis.py [hunk 1]
```diff
@@ -463,6 +463,8 @@ def get_best_checkpoint(
                 "training_iteration" is used by default if no value was
                 passed to ``self.default_metric``.
             mode: One of [min, max]. Defaults to ``self.default_mode``.
+            return_path: If True, only returns the path (and not the
+                ``Checkpoint`` object).
 
         Returns:
             :class:`Checkpoint <ray.air.Checkpoint>` object.
```
Comment (xwjiang2010) [inline python/ray/tune/analysis/experiment_analysis.py:470]: may be a path as well?

File: python/ray/tune/trainable/trainable.py [hunk 2]
```diff
@@ -668,9 +674,9 @@ def delete_checkpoint(self, checkpoint_path: str):
         Args:
             checkpoint_path: Path to checkpoint.
```
Comment (xwjiang2010) [inline python/ray/tune/trainable/trainable.py:675]: update type hints/docs?
no long str anymore?

**Review by xwjiang2010 — APPROVED** [review]
Thanks!
Let's think about how to present the checkpoint on client side.
I am also thinking that we should add some documentation around this concept. It's not obvious for first time users.

**Pull request merged**